packages/util-endpoints/src/resolveEndpoint.ts

```diff
@@ -0,0 +1,33 @@
+import { EndpointError, EndpointResolverOptions, RuleSetObject } from "./types";
+import { evaluateRules } from "./utils";
+
+/**
+ * Resolves an endpoint URL by processing the endpoints ruleset and options.
+ */
+export const resolveEndpoint = (ruleSetObject: RuleSetObject, options: EndpointResolverOptions) => {
+  const { endpointParams, logger } = options;
+  const { parameters, rules } = ruleSetObject;
+
+  const requiredParams = Object.entries(parameters)
+    .filter(([, v]) => v.required)
+    .map(([k]) => k);
+
+  for (const requiredParam of requiredParams) {
+    if (endpointParams[requiredParam] == null) {
+      throw new EndpointError(`Missing required parameter: '${requiredParam}'`);
+    }
+  }
+
+  // @ts-ignore Type 'undefined' is not assignable to type 'string | boolean' (2322)
+  const paramsWithDefault: [string, string | boolean][] = Object.entries(parameters)
+    .filter(([, v]) => v.default != null)
+    .map(([k, v]) => [k, v.default]);
+
+  if (paramsWithDefault.length > 0) {
+    for (const [paramKey, paramDefaultValue] of paramsWithDefault) {
+      endpointParams[paramKey] = endpointParams[paramKey] ?? paramDefaultValue;
+    }
+  }
+
+  return evaluateRules(rules, { endpointParams, logger, referenceRecord: {} });
```

Comment:
What's the difference of `endpointParams` and `referenceRecord`? Can they merge into a simple `endpointParamsAndReference`?

Reply:
I see a few places below the 2 are merged, maybe they can be simplified.

Reply:
The `referenceRecord` is valid only per conditions block. The `endpointParams` are valid across the resolution.
They can be merged, but that would increase the complexity. We'll have to manage keys of references in a different data structure, and remove those keys from `endpointParamsAndReference` before the peer conditions block is processed.

Reply:
For the sake of simplicity, I'll retain the `endpointParams` and `referenceRecord` options.